README.md
<!-- markdownlint-disable MD033 -->
# LIFX Day and Dusk in a Container

[![GitHub issues](https://img.shields.io/github/issues/djelibeybi/docker-lifx-daydusk?logo=github&style=for-the-badge)](https://github.com/Djelibeybi/docker-lifx-daydusk/issues)
[![Docker Pulls](https://img.shields.io/docker/pulls/djelibeybi/lifx-daydusk?logo=docker&style=for-the-badge)](https://hub.docker.com/r/djelibeybi/lifx-daydusk)

This container image reproduces the LIFX Day and Dusk scheduling functionality locally but removes the dependency on the LIFX Cloud and adds fine-grained control over bulb selection, timing, kelvin value and power status.

![Alt](https://repobeats.axiom.co/api/embed/d94f12c0421324e280e6303aa3cb4e01fb401943.svg "Repobeats analytics image")

## Supported Platforms

This image is supported on the following platforms:

* `amd64`: 64-bit Intel or AMD processors including Intel-based Synology NAS  devices.
* `aarch64`: 64-bit Arm processors including Raspberry Pi 3 Model B/B+ and  Raspberry Pi 4.
* `aarch32`: 32-bit build made for Raspian which defaults to a 32-bit install.

This image is supported with the following OS/runtime combinations:

* **macOS**: Docker Desktop
* **Windows 10/11**: Docker Desktop using Windows Subsystem for Linux
* **Linux**: Docker CE and Podman

## Usage

LIFX discovery relies on UDP broadcasts on the local network in order to find bulbs. This can be achieved using the ```--net=host``` parameter on Linux, but this does **not work** when using [Docker Desktop for Mac](https://github.com/docker/for-mac/issues/68) or [Docker Desktop for Windows](https://github.com/docker/for-win/issues/543).

However, as UDP broadcasts are only required for discovery, you can workaround this by using [`hardcoded_discovery`](https://photons.delfick.com/configuration/discovery_options.html#discovery-options) instead, which only requires unicast UDP support and that works on all platforms. Hardcoded discovery can also be used to traverse VLANs.

The following command will start the container with the required mandatory configuration:

```bash
docker run \
  --detach \
  --name=daydusk \
  --net=host \ # Not required if hardcoded_discovery is used
  -e TZ=<Time Zone> \
  -v /path/to/config/:/config/ \
  djelibeybi/lifx-daydusk
```

### Parameters

The following parameters can be specified on the command-line when executing  `docker run` to start the container:

* `-e TZ`: for [timezone information](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones) e.g. `-e TZ=Europe/London` (**required**)
* `-v /path/to/config/:/config/` - **Required:** see below for details.
* `--net=host`: This is **required** to support automatic discovery but only works on Linux. On other platforms, use `hardcoded_discovery` via the `lifx.yml` configuration file instead of using this parameter.

> **IMPORTANT:** You **must** set a valid `TZ` value otherwise the timezone inside the container will be set to UTC and your events will fire at the wrong time.

The `--net=host` option can be replaced by advanced users with an appropriate `macvlan` network configuration however the configuration of the network itself and modification of the `docker run` command to use that network is left as an exercise for the reader. For most users, using `hardcoded_discovery` is far easier.

## Creating the `daydusk.yml` configuration file

The sample `docker run` command above maps the local `/path/to/config` directory to the `/config` directory inside the container. You should change `/path/to/config` to an actual directory on your host system and you must create at least the `daydusk.yml` file in this directory.

> **NOTE:** the [`sample-daydusk.yml`](https://github.com/Djelibeybi/docker-lifx-daydusk/blob/develop/sample-daydusk.yml) matches the default LIFX Day & Dusk configuration **exactly** including powering on the specified bulbs at the start of each event.

### Syntax

The `daydusk.yml` file consists of one or more named events that define the time  to start the transition, the end state of the bulbs, the duration of the  transition and whether the bulbs should turn on automatically before the  transition starts or turn off automatically after the transition ends.

In the example below, the event named `wakeup` will fire at **6:30am** on Saturday and Sunday to trigger the bulbs to power on _in whatever state they were in when they were previously powered off_ and then begin a **30 minute** transition to **80% brightness** at **4000 kelvin**:

```yaml
daydusk:
  schedules:
    wakeup:
      days:
        - SATURDAY
        - SUNDAY
      hour: 6
      minute: 30
      brightness: 0.8
      kelvin: 4000
      duration: 1800
      power: ON
      transform_options:
        transition_color: True
```

Multiple events can be included within the file, using the following syntax:

```yaml
daydusk:
  schedules:
     wakeup:
     ...

     day:
     ...

     evening:
     ...

     nighttime:
     ...
```

Note the opening `daydusk:` and `schedules:` only appear once at the beginning of the file.

> **Tip:** [You can validate your YAML syntax online](http://www.yamllint.com/).

There is no limit to the number of events, though each event will add additional time to process when the container starts. To change the events, use `docker stop` to shut down the running container, modify the `daydusk.yml` file and then run `docker start` to start the container back up again.

> **Note:** The container will rebuild the schedule from `daydusk.yml` at startup.

The following table documents each parameter and all parameters are required for each event:

| Parameter    | Required? | Value            | Detail |
| ------------ | :-------- | :--------------- | :----- |
| `days`       | No        | List of either `MONDAY` to `SUNDAY` or `0` to `6` | An list of days on which this event should occur. If not specified, the event will occur daily.<br>If specified numerically, `0` is Sunday and `6` is Saturday. |
| `hour`       | **Yes**   | `0` to `23`     | The hour at which the transition starts |
| `minute`     | **Yes**   |`0` to `59`      | The minute after the hour at which the transition starts |
| `task`       | No        | `theme`         | If you want to apply a theme to the selected bulbs, set this to `theme`. |
| `hue`        | No        |`0` to `360`     | The target hue in degrees at the end of the transition.<br>**Must be set to `0` for temperature adjustment** |
| `saturation` | No        |`0` to `1`       | The target saturation as a float, e.g. for 80% specify `0.8`.<br>**Must be set to `0` for temperature adjustment** |
| `brightness` | **Yes**   |`1` to `1`       | The target brightness as a float, e.g. for 80% specify `0.8` |
| `kelvin`     | **Yes**   |`1500` to `9000` | The target kelvin value at the end of the transition.<br>**Ignored unless both `hue` and `saturation` are set to `0`** |
| `colors`     | No        | -               | An optional [list of colors](#specifying-colors-for-a-theme) to use when applying a theme. Only valid if `task` is set to `theme`. |
| `duration`   | **Yes**   |`1` to `86400`   | How long the transition should run in seconds.<br>Sixty minutes is `3600` seconds. |
| `power`      | No        | <code>[on&#124;off]</code> | Used to turn the bulbs `on` at the start of the event or `off` when the event ends. If not provided the power state remains unchanged. |
| `transform_options` | No | -               | An optional [list of options](#adding-options-for-each-event) to apply to each event. |
| `override`   | No        | -               | An optional [list of overrides](#adding-overrides-for-a-theme) to apply when applying a theme. |
| `reference`  | No        | -               | Used to [specify the bulbs](#specifying-bulbs-for-each-event) to target for each event. |

The [`sample-daydusk.yml`](sample-daydusk.yml) file contains four events that replicate the default LIFX Day & Dusk times, brightness and kelvin values as well as the transition duration and power state changes.

The [`theme-daydusk.yml`](theme-daydusk.yml) will set a rainbow theme across all devices.

### Adding options for each event

There are two options that can be used to fine-tune the transformation that occurs for each event:

* `transition_color`: modifies the behaviour of the bulb if it's powered on by the event. By default if `power` is set to `on`, the target bulb(s) will be set to the target HSBK or color value before being powered on and the transition duration will only affect the brightness. By setting this to `True`, the target bulb(s) will power on with the existing HSBK values and the transition will affect all values, i.e. the bulb will transition both color and brightness over  the duration. This is particularly useful for `wakeup` transitions so that the  initial power on doesn't jump to a much higher kelvin value immediately.
* `keep_brightness`: modifies the transition to ignore any brightness value, i.e. only the HSK/color values are transitions over the duration while the brightness remains the same.

Example usage:

```yaml
transform_options:
  transition_color: True
  keep_brightness: False
```

### Specifying colors for a theme

The `colors` parameter is optional. If omitted, the default theme uses seven colors at 30% brightness.

To specify your own colors, use any combination of the following formats:

* Names: `white`, `red`, `orange`, `yellow`, `cyan`, `green`, `blue` `purple`, `pink`
* Hexadecimal: `hex:#RRGGBB`
* Red, green, blue: `rgb:0-255,0-255,0-255`
* Hue, saturation and brightness: `hsb:0-359, 0-1, 0-1`

Here is an example list of colors that uses each of these formats:

```yaml
colors:
  - red
  - hex:#00ff00
  - rgb:0, 0, 255
  - hsb:120, 1, 0.7
```

Only the `hsb` option specifies brightness for each color. You can override the brightness globally using the `override` option.

### Adding overrides for a theme

By default, Photons will use the current brightness of the target device(s) when applying a theme with specified colors that don't include a specific brightness value. This can be overridden by providing a brightness for all devices when applying the theme.

To specify a global brightness, add it as an override to your event. The following example sets the brightness to 70%:

```yaml
override:
  - brightness: 0.7
```

### Specifying bulbs for each event

The `reference` field is used to determine which bulbs will be targeted for each event. If an event does not contain a reference field, all discovered bulbs will be used. The reference field can be specified in any of the following formats on a per-event basis, i.e. you can chose to use one method for an event  and a different method for another.

> **No validation is performed to ensure the serial numbers are valid or if any specified files exist.** Likewise, no validation of the provided filters is performed, so it's possible for a filter to return no results or unexpected results if misconfigured.

The available formats are:

**A single serial number:**

```yaml
reference: d073d5000001
```

**A list of serial numbers:**

```yaml
reference:
  - d073d5000001
  - d073d5000002
```

**A file that contains a list of serial numbers (one per line):**

```yaml
reference: file:/config/bulbs.conf
```

**A filter that is dynamically evaluated on each run based on bulb-specific data:**

```yaml
reference: match:group_name=bedroom
```

The Photons documentation maintains a list [valid filters](https://photons.delfick.com/commandline/references.html#the-match-reference) that can be used with this option. Multiple filters can be combined using an ampersand, e.g. `match:group_name=bedroom&power=on` would match all bulbs in the group named `bedroom` that are currently powered on.

### Additional Photons configuration

To provide [additional configuration options](https://photons.delfick.com/configuration/index.html), including `hardcoded_discovery` entries, place them in a file named `lifx.yml` file in the same directory as your `daydusk.yml` file.

## Acknowledgements

* [delfick](https://github.com/delfick) for [Photons](https://photons.delfick.com) upon which my code depends and for providing additional pointers on usingPhotons for configuration validation.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format
is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) and this
project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2.4.1

- Minor version release to update the image:
  - Updated base image to `python-3.10.4-slim`
  - Updated to [Photons 0.42.3](https://photons.delfick.com/changelog.html)

## 2.4.0

- Minor version release to update the image:
  - Updated base image to `python-3.10.0-slim`
  - Updated Photons to 0.41.0

## 2.3.3

- Update Supercronic to v0.1.12
- Switch to `python:3.9.1-slim` so that Dependabot alerts me on a new release.

## 2.3.2

- Provide more logging in the case of schedule failures

## 2.3.1

- Container image size optimisation. Reduced the final image size by 100MB
- Build/test CI changed from Travis to GitHub Actions.

## 2.2.1

- Rebuild release: triggered to update the container image on Docker Hub

## 2.2.0

### Added in v2.2.0

- **Themes:** you can now apply a theme to your devices instead of a transition.

### Fixed in v2.2.0

- Recreate crontab on startup instead of reading existing crontab file.

### Changed in v2.2.0

- Updated to [Photons](https://github.com/delfick/photons) v0.30.0
- Switched from S6-overlay to Supercronic to reduce image size and memory
consumption

## 2.1.0

### Added in v2.1.0

- Added the ability to set `transform_options` for each schedule.

### Changed in v2.1.0

- Minor formatting changes to various Markdown files to fix linting issues.

## 2.0.0

> **BREAKING CHANGES!**
> This release contains **breaking changes** since the 1.1.0 release. An
existing `daydusk.json` configuration will no longer work.

- Switched to using YAML configuration to take advantage of the underlying
Photons library validation capabilities
- A new `sample-daydusk.yml` is provided that matches the default Day & Dusk
funtionality provided by LIFX
- Updated documentation for the new YAML format
- Bulbs can be specified on a per event basis in a far more flexible manner
either within the configuration or via external files or filters.

### Added in v2.0.0

- Hue and saturation can be set for each event along with brightness and kelvin.
  Note that if hue and/or saturation are set to any value greater than `0`,
  the kelvin value is ignored.
- Events can be limited to specific days or will run on every day (default)

### Changed in v2.0.0

- Logging configuration modified so that only LIFX-related errors are logged.

## 1.1.0

### Added in v1.1.0

- Bulbs can now be targeted per event using `<event>-bulbs.conf` files.

### Removed in v1.1.0

- A work-in-progress script was committed accidentally and has been removed.

## 1.0.0

Initial stable release.

### Added in v1.0.0

- Basic LIFX Day & Dusk functionality
- Ability to specify target bulbs
- Ability to configure each transition, including power state
- Multi-arch image build support via Travis CI
- Manifest support on Docker Hub


rootfs/scripts/generate-crontab.py
#!/usr/local/bin/python3
"""
A script for generating a crontab that can be used to simulate

It takes in options from photons configuration that looks a bit like:

    ---

    daydusk:

      schedules:
        day:
          hour: 7
          minute: 0
          days:
            - MONDAY
            - WEDNESDAY
          hue: 0
          brightness: 0.9
          kelvin: 3500
          duration: 60
          power: ON

        evening:
          hour: 0
          minute: 0
          brightness: 0.8
          kelvin: 2700
          duration: 600
          power: ON

The options you have available for each schedule is:

hue - integer between 0 and 360 - default 0
saturation - float between 0 and 1 - default 0
brightness - float between 0 and 1 - default 1
kelvin - integer between 1500 and 9000 - default 3500
power - ON/OFF, true/false; or 0/1
duration - A unbounded float of the number of seconds you want this to take

days
    The days you want this run. If not specified it'll run on all days.
    Otherwise this should be a list of the days. Each day can be the name of
    the day (i.e. MONDAY, WEDNESDAY, ...) or a number from 0 to 6 where sunday
    is 0.


hour - required integer between 0 and 23
minute - required integer between 0 and 59

reference
    The lights to target. If this is an empty string or an underscore or not
    specified then it'll default to all devices that can be found.

    Otherwise it should be a special reference string. For example:
    ``match:label=kitchen`` or ``file:/path/to/file/with_mac_per_line``

    Or a list of device mac addresses (i.e. d073d5001337)
"""

from textwrap import dedent
import distutils.sysconfig
import logging
import shlex
import json
import enum
import os

from crontab import CronTab

from photons_app.formatter import MergedOptionStringFormatter
from photons_app.errors import PhotonsAppError
from photons_app.actions import an_action

from photons_protocol.types import enum_spec

from photons_control.colour import make_hsbk

from delfick_project.norms import dictobj, sb, Meta, BadSpecValue
from delfick_project.addons import addon_hook

log = logging.getLogger("daydusk")

default_colors = [
    (0, 1, 0.3, 3500),
    (40, 1, 0.3, 3500),
    (60, 1, 0.3, 3500),
    (127, 1, 0.3, 3500),
    (239, 1, 0.3, 3500),
    (271, 1, 0.3, 3500),
    (294, 1, 0.3, 3500),
]

class NoSchedules(PhotonsAppError):
    desc = dedent(
        """
    You need to specify some schedules under the ``daydusk.schedules`` section
    of your configuration

    For example::

        ---

        daydusk:
          schedules:
            day:
              hour: 7
              minute: 0
              days:
                - MONDAY
                - WEDNESDAY

              hue: 0
              saturation: 0
              brightness: 0.9
              kelvin: 3500
              duration: 60
              power: ON
              transform_options:
                - transition_color: True
    """
    ).rstrip()

class Days(enum.Enum):
    SUNDAY = 0
    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5
    SATURDAY = 6

class range_spec(sb.Spec):
    def setup(self, spec, minimum, maximum):
        self.spec = spec
        self.minimum = minimum
        self.maximum = maximum

    def normalise_filled(self, meta, val):
        val = self.spec.normalise(meta, val)
        if val < self.minimum or val > self.maximum:
            raise BadSpecValue(
                "Value must be between min and max values",
                wanted=val,
                minimum=self.minimum,
                maximum=self.maximum,
                meta=meta,
            )
        return val

class power_spec(sb.Spec):
    def normalise_filled(self, meta, val):
        if val in ("on", "ON", True, 1):
            return "on"
        elif val in ("off", "OFF", False, 0):
            return "off"

        raise BadSpecValue("Power must be on/off, True/False or 0/1", wanted=val, meta=meta)

class task_spec(sb.Spec):
    def normalise_empty(self, meta):
        return "lan:transform"

    def normalise_filled(self, meta, val):
        if val in ("theme"):
            return "lan:apply_theme"
        else:
            return "lan:transform"

class colors_spec(sb.Spec):
    def normalise_empty(self, meta):
        return default_colors

    def normalise_filled(self, meta, val):
        cs = [make_hsbk(val) for val in val]
        return [(c["hue"], c["saturation"], c["brightness"], c["kelvin"]) for c in cs]

class reference_spec(sb.Spec):
    def normalise_empty(self, meta):
        return ""

    def normalise_filled(self, meta, val):
        return ",".join(sb.listof(sb.string_spec()).normalise(meta, val))

class Schedule(dictobj.Spec):
    days = dictobj.NullableField(sb.listof(enum_spec(None, Days, unpacking=True)))
    hour = dictobj.Field(range_spec(sb.integer_spec(), 0, 23), wrapper=sb.required)
    minute = dictobj.Field(range_spec(sb.integer_spec(), 0, 59), wrapper=sb.required)

    task = dictobj.Field(task_spec)

    hue = dictobj.Field(range_spec(sb.float_spec(), 0, 360), default=0)
    saturation = dictobj.Field(range_spec(sb.float_spec(), 0, 1), default=0)
    brightness = dictobj.Field(range_spec(sb.float_spec(), 0, 1), default=1)
    kelvin = dictobj.Field(range_spec(sb.integer_spec(), 1500, 9000), default=3500)
    transform_options = dictobj.NullableField(sb.dictof(sb.string_spec(), sb.boolean()))

    duration = dictobj.NullableField(sb.float_spec)
    power = dictobj.NullableField(power_spec)

    colors = dictobj.NullableField(colors_spec)
    override = dictobj.NullableField(sb.dictof(sb.string_spec(), range_spec(sb.float_spec(), 0, 1)))

    reference = dictobj.Field(reference_spec)

    @property
    def hsbk(self):
        if self.task == 'lan:transform':
            keys = ["hue", "saturation", "brightness", "kelvin"]
            options = {k: v for k, v in self.as_dict().items() if k in keys}
            return {k: v for k, v in options.items() if v is not None}
        else:
            return {}

    @property
    def extra(self):
        keys_except = ["days", "hour", "minute", "reference", "task", "hue", "saturation", "brightness", "kelvin"]
        options = {k: v for k, v in self.as_dict().items() if k not in keys_except}
        return {k: v for k, v in options.items() if v is not None}

    @property
    def dow(self):
        days = self.days
        if not self.days:
            days = list(Days)

        return [day.value for day in days]

class DayDusk(dictobj.Spec):
    schedules = dictobj.Field(
        sb.dictof(sb.string_spec(), Schedule.FieldSpec(formatter=MergedOptionStringFormatter))
    )

@an_action(needs_target=False, special_reference=False)
async def make_crontab(collector, target, reference, artifact, **kwargs):
    """
    Make a crontab file executing our day dusk options.

    Usage is::

        ./generate-crontab.py
    """
    collector.register_converters(
        {"daydusk": DayDusk.FieldSpec(formatter=MergedOptionStringFormatter)}
    )
    daydusk = collector.configuration["daydusk"]

    spec = sb.set_options(
        path=sb.defaulted(sb.string_spec(), '/config/daydusk.crontab'),
        lifx_script=sb.defaulted(sb.string_spec(), '/usr/local/bin/lifx')
    )
    extra = collector.configuration["photons_app"].extra_as_json
    kwargs = {
        k: v for k, v in spec.normalise(Meta.empty(), extra).items() if v is not sb.NotSpecified
    }

    cronfile = kwargs['path']
    lifx_script = kwargs['lifx_script']

    if not daydusk.schedules:
        raise NoSchedules()

    cron = CronTab()

    extra_script_args = ["--silent"]

    for name, options in daydusk.schedules.items():
        script_args = {**options.hsbk, **options.extra}
        command = [
            lifx_script,
            options.task,
            options.reference,
            *extra_script_args,
            "--",
            json.dumps(script_args),
        ]

        command = str(" ".join([shlex.quote(part) for part in command]))

        job = cron.new(command=command)
        job.dow.on(*options.dow)
        job.minute.on(options.minute)
        job.hour.on(options.hour)

    if os.path.exists(cronfile):
        os.remove(cronfile)

    cron.write(cronfile)
    print(f"Generated crontab at {cronfile}")

if __name__ == "__main__":
    __import__("photons_core").run('lan:make_crontab --silent {@:1:}')
